third_party/move/move-model/src/builder/exp_builder.rs

```diff
@@ -4136,14 +4136,57 @@ impl<'env, 'translator, 'module_translator> ExpTranslator<'env, 'translator, 'mo
         context: &ErrorMessageContext,
     ) -> ExpData {
         // Translate arguments.
-        let (arg_types, translated_args) = self.translate_exp_list(args);
+        let (mut arg_types, mut translated_args) = self.translate_exp_list(args);
 
         // Special handling of receiver call functions
         if kind == CallKind::Receiver {
             debug_assert!(
                 module.is_none(),
                 "unexpected qualified name in receiver call"
             );
+            debug_assert!(
+                !args.is_empty(),
+                "receiver call needs to have at least one parameter"
+            );
+            let receiver_call_opt = self.get_receiver_function(&arg_types[0], name);
+            if let Some(receiver_call) = receiver_call_opt {
+                if let Exp_::ExpDotted(dotted) = &args[0].value {
+                    // special case for the receiver call S[x].f.fun(&mut...)
```

Comment:
Would be useful to add some note on why these special cases are needed (e.g., we resolve X before doing Y translation), for future refactors/bug fixes.

Reply:
done